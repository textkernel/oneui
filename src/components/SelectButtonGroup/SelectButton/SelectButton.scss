.SelectButton {
    box-sizing: border-box;
    border: var(--line-normal) solid transparent;
    padding: var(--spacing-2x) var(--spacing-4x);
    line-height: var(--line-height-normal);
    white-space: nowrap;
    text-overflow: ellipsis;
    text-align: center;
    overflow: hidden;
    cursor: pointer;
    outline: none;
    color: var(--color-neutral-75);
    font-weight: var(--font-weight-bold);

    &:first-child {
        border-top-left-radius: var(--border-radius);
        border-bottom-left-radius: var(--border-radius);
    }

    &:last-child {
        border-top-right-radius: var(--border-radius);
        border-bottom-right-radius: var(--border-radius);
    }

    &:hover {
        background: var(--color-light);
    }

    + .SelectButton {
        box-shadow: -1px 0 0 0 var(--color-background);
    }

    &[disabled] {
        cursor: not-allowed;
    }

    &--isSelected.SelectButton {
        border-color: var(--color-primary);
        background-color: var(--color-primary);
        color: var(--color-background);
        font-weight: var(--font-weight-bold);

        &:hover {
            border-color: var(--color-primary-60);
            background-color: var(--color-primary-60);
        }

    }

    &--size {
        &_small {
            font-size: var(--font-size-small);
            line-height: var(--line-height-small);
            padding: var(--spacing-normal) var(--spacing-2x);
        }
        &_large {
            font-size: var(--font-size-large);
            line-height: var(--line-height-large);
            padding: var(--spacing-3x) var(--spacing-5x);
        }
    }

    &--isEqualWidth {
        flex-grow: 1;
        flex-basis: 0;
    }

    &--isBlock {
        flex-grow: 1;
    }
}
