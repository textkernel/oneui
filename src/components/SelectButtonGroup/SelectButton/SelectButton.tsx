import * as React from 'react';
import { bem } from '../../../utils';
import { ENTER_KEY, Size } from '../../../constants';
import styles from './SelectButton.scss';
import { NotEmptyReactNode } from '../../../customTypes/types';

// These props will be passed by the parent <SelectButtonGroup>
interface InternalProps<V> extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange'> {
    /** A function to be called if a button is pressed. */
    onChange?: (value: V) => void;
    /** If this component a child of a full block with container  */
    isBlock?: boolean;
    /** If this button has same width as its siblings */
    isEqualWidth?: boolean;
}

export interface Props<V> extends InternalProps<V> {
    /** the label to be displayed on this button */
    children: NotEmptyReactNode;
    /** the value associated with this button */
    value: V;
    /** whether or not this button is selected */
    isSelected?: boolean;
    /** size of the button */
    size?: Size;
}

const { block } = bem('SelectButton', styles);

export function SelectButton<V>({
    children,
    value,
    onChange,
    isEqualWidth = false,
    isBlock,
    isSelected = false,
    size = 'normal',
    ...rest
}: Props<V>) {
    const handleClick = () => {
        if (onChange) {
            onChange(value);
        }
    };

    const handleKeyPress = (e: React.KeyboardEvent<HTMLDivElement>) => {
        if (e.key === ENTER_KEY) {
            handleClick();
        }
    };

    return (
        <div
            role="button"
            tabIndex={0}
            {...rest}
            onKeyPress={handleKeyPress}
            onClick={handleClick}
            {...block({ isEqualWidth, isSelected, size, ...rest })}
        >
            {children}
        </div>
    );
}

SelectButton.displayName = 'SelectButton';
