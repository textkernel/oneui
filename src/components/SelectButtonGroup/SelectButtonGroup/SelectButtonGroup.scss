.SelectButtonGroup {
    background-color: var(--color-neutral-10);
    display: inline-flex;
    border-radius: var(--border-radius);

    &--isBlock {
        display: flex;
        width: 100%;
    }
}
