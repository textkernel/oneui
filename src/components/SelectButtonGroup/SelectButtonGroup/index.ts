export { SelectButtonGroup, Props as SelectButtonGroupProps } from './SelectButtonGroup';
