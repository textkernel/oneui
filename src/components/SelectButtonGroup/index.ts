export { SelectButton, SelectButtonProps } from './SelectButton';
export { SelectButtonGroup, SelectButtonGroupProps } from './SelectButtonGroup';
