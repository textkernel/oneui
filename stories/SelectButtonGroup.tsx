import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { text, boolean, select, withKnobs } from '@storybook/addon-knobs';
import { SelectButtonGroup, SelectButton } from '@textkernel/oneui';
import { CONTEXTS, SIZES } from '../src/constants';

storiesOf('Atoms|SelectButtonGroup', module)
    .addDecorator(withKnobs)
    .add('Uncontrolled', () => (
        <SelectButtonGroup<string>
            isMultiselect={boolean('Multiselect group', false)}
            isRequired={boolean('Require to have at least 1 selected option', false)}
            isBlock={boolean('Display as block', false)}
            isEqualWidth={boolean('Make all buttons the same width', false)}
            context={select('Default selected color context', CONTEXTS, CONTEXTS[1])}
            size={select('Size', SIZES, SIZES[1])}
            defaultValue={['3']}
            onChange={(selection) => {
                console.log('Current selection', selection);
            }}
        >
            <SelectButton value="1" key="button 1">
                {text('Option label 1', 'Option 1')}
            </SelectButton>
            <SelectButton value="2" key="button 2">
                {text('Option label 2', 'Option 2')}
            </SelectButton>
            <SelectButton value="3" key="button 3">
                {text('Option label 3', 'Option 3')}
            </SelectButton>
        </SelectButtonGroup>
    ))
    .add('Controlled', () => {
        const buttons = {
            '1': boolean('Button 1 selected', true),
            '2': boolean('Button 2 selected', false),
            '3': boolean('Button 3 selected', false),
        };
        const keys = Object.keys(buttons);

        return (
            <SelectButtonGroup
                value={keys.filter((k) => buttons[k])}
                onChange={(value) => {
                    console.log(`Button ${value} was selected`);
                }}
            >
                {keys.map((buttonNumber) => (
                    <SelectButton
                        key={`button-${buttonNumber}`}
                        value={buttonNumber}
                        context={select(
                            `Context for button ${buttonNumber}`,
                            CONTEXTS,
                            CONTEXTS[1]
                        )}
                    >
                        Button {buttonNumber}
                    </SelectButton>
                ))}
            </SelectButtonGroup>
        );
    });
