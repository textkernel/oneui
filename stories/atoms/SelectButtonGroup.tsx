import * as React from 'react';
import { SelectButtonGroup, SelectButton } from '@textkernel/oneui';

export default {
    title: 'Atoms/SelectButtonGroup',
    component: SelectButtonGroup,
    subcomponents: { SelectButton },
};

export const Uncontrolled = (args) => (
    <SelectButtonGroup<string> {...args}>
        <SelectButton value="1" key="button 1">
            Option 1
        </SelectButton>
        <SelectButton value="2" key="button 2">
            Option 2
        </SelectButton>
        <SelectButton value="3" key="button 3">
            Option 3
        </SelectButton>
    </SelectButtonGroup>
);
Uncontrolled.args = {
    defaultValue: ['3'],
};

export const Controlled = (args) => {
    const buttons = {
        '1': args.button1,
        '2': args.button2,
        '3': args.button3,
    };
    const keys = Object.keys(buttons);

    return (
        <SelectButtonGroup value={keys.filter((k) => buttons[k])}>
            {keys.map((buttonNumber) => (
                <SelectButton key={`button-${buttonNumber}`} value={buttonNumber}>
                    Button {buttonNumber}
                </SelectButton>
            ))}
        </SelectButtonGroup>
    );
};
Controlled.args = {
    button1: true,
    button2: false,
    button3: false,
};
