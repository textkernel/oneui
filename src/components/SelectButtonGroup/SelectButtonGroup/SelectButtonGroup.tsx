import * as React from 'react';
import { bem } from '../../../utils';
import { Context, Size } from '../../../constants';
import { SelectButtonProps } from '../SelectButton';
import styles from './SelectButtonGroup.scss';

export interface Props<V>
    extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange' | 'defaultValue'> {
    /** SelectButton children */
    children: React.ReactElement<SelectButtonProps<V>>[];
    /** Function that is called when the selection is changed (controlled use) */
    onChange?: (selection: V | V[]) => void;
    /** should more than one option be allowed to be selected (uncontrolled use) */
    isMultiselect?: boolean;
    /** if required, there should at least be one selected value (uncontrolled use) */
    isRequired?: boolean;
    /** should the component take up all the width available */
    isBlock?: boolean;
    /** Color context for selected buttons */
    context?: Context;
    /** should children have equal width */
    isEqualWidth?: boolean;
    /** size of the button group */
    size?: Size;
    /** currently selected value(s) (controlled use) */
    value?: V[];
    /** currently selected value(s) (uncontrolled use) */
    defaultValue?: V[];
}

const { block } = bem('SelectButtonGroup', styles);

export function SelectButtonGroup<V>(props: Props<V>) {
    const {
        children,
        isMultiselect,
        isRequired,
        isEqualWidth,
        isBlock,
        context,
        onChange,
        size,
        value,
        defaultValue,
        ...rest
    } = props;

    const [selection, setSelection] = React.useState(defaultValue || []);

    const handleChange = (selectedValue: V) => {
        if (value) {
            // If value is present, this means state is controlled outside of the component
            // Only the currently selected value will be passed as argument
            if (onChange) {
                onChange(selectedValue);
            }
        } else {
            // Uncontrolled use: state is managed internally
            const isCurrentlySelected = selection.includes(selectedValue);
            let newSelection = selection;

            if (!isMultiselect) {
                if (isRequired || !isCurrentlySelected) {
                    // Select a value other than the one currently selected
                    newSelection = [selectedValue];
                } else {
                    // Deselect the currently selected value
                    newSelection = [];
                }
            } else if (isCurrentlySelected) {
                if (!isRequired || selection.length > 1) {
                    // If some selection is required, only allow deselection
                    // if there are at least two values selected
                    newSelection = [...selection.filter((v) => v !== selectedValue)];
                }
            } else {
                // Add value to existing selection
                newSelection = [...selection, selectedValue];
            }

            setSelection(newSelection);

            if (onChange) {
                // Communicate new selection to the outside
                onChange(newSelection);
            }
        }
    };

    return (
        <div {...rest} {...block(props)}>
            {children.map((child) => {
                const childProps = {
                    context: context || child.props.context,
                    isBlock,
                    isEqualWidth,
                    isSelected:
                        (value || selection).includes(child.props.value) || child.props.isSelected,
                    onChange: handleChange,
                    size,
                };

                return React.cloneElement(child, childProps);
            })}
        </div>
    );
}

SelectButtonGroup.displayName = 'SelectButtonGroup';

SelectButtonGroup.defaultProps = {
    isMultiselect: false,
    isRequired: false,
    isBlock: false,
    isEqualWidth: false,
    context: null,
    size: 'normal',
    onChange: null,
    value: null,
    defaultValue: null,
};
