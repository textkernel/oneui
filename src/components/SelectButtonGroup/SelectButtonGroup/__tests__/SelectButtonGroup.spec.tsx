import * as React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { SelectButtonGroup } from '../SelectButtonGroup';
import { SelectButton } from '../../SelectButton';

describe('SelectButtonGroup', () => {
    let view: RenderResult;
    const onChangeMock = jest.fn();
    const defaultProps = {
        defaultValue: ['1'],
        onChange: onChangeMock,
    };
    const rerenderView = (props) => {
        view.rerender(
            <SelectButtonGroup {...defaultProps} {...props}>
                <SelectButton value="1" key="1">
                    Option 1
                </SelectButton>
                <SelectButton value="2" key="2">
                    Option 2
                </SelectButton>
                <SelectButton value="3" key="3">
                    Option 3
                </SelectButton>
            </SelectButtonGroup>
        );
    };
    const getButton = (number) => screen.getAllByRole('button')[number];

    beforeEach(() => {
        view = render(
            <SelectButtonGroup {...defaultProps}>
                <SelectButton value="1" key="1">
                    Option 1
                </SelectButton>
                <SelectButton value="2" key="2">
                    Option 2
                </SelectButton>
                <SelectButton value="3" key="3">
                    Option 3
                </SelectButton>
            </SelectButtonGroup>
        );
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    it('should render correctly', () => {
        expect(view.container).toMatchSnapshot();
    });

    describe('Single select mode', () => {
        it('should not change isSelect prop of children on initial render', () => {
            expect(getButton(0)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).not.toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).not.toHaveClass('SelectButton SelectButton--isSelected');
        });

        it('should switch the selected option if a not-selected option is clicked', async () => {
            const user = userEvent.setup();
            await user.click(getButton(1));

            expect(getButton(0)).not.toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).not.toHaveClass('SelectButton SelectButton--isSelected');
        });

        it('should allow empty selection if field is not required', async () => {
            const user = userEvent.setup();
            await user.click(getButton(0));

            expect(getButton(0)).not.toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).not.toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).not.toHaveClass('SelectButton SelectButton--isSelected');
        });

        it('should not allow empty selection if field is required', async () => {
            const newProps = {
                isRequired: true,
            };
            rerenderView(newProps);
            const user = userEvent.setup();
            await user.click(getButton(0));

            expect(getButton(0)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).not.toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).not.toHaveClass('SelectButton SelectButton--isSelected');
        });

        it('should call onChange with correct parameters', async () => {
            const user = userEvent.setup();
            await user.click(getButton(1));

            expect(onChangeMock).toHaveBeenCalledWith(['2']);

            await user.click(getButton(0));

            expect(onChangeMock).toHaveBeenLastCalledWith(['1']);

            await user.click(getButton(0));

            expect(onChangeMock).toHaveBeenLastCalledWith([]);
        });
    });

    describe('Multi select mode', () => {
        beforeEach(() => {
            const newProps = {
                isMultiselect: true,
            };
            rerenderView(newProps);
        });

        it('should not change isSelect prop of children on initial render', () => {
            expect(getButton(0)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).not.toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).not.toHaveClass('SelectButton SelectButton--isSelected');
        });

        it('should switch to selected if a not-selected option is clicked, without changing others', async () => {
            const user = userEvent.setup();
            await user.click(getButton(1));

            expect(getButton(0)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).not.toHaveClass('SelectButton SelectButton--isSelected');

            await user.click(getButton(2));

            expect(getButton(0)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).toHaveClass('SelectButton SelectButton--isSelected');
        });

        it('should switch to not-selected if a selected option is clicked, without changing others', async () => {
            const user = userEvent.setup();
            await user.click(getButton(1));

            expect(getButton(0)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).not.toHaveClass('SelectButton SelectButton--isSelected');

            await user.click(getButton(1));

            expect(getButton(0)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).not.toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).not.toHaveClass('SelectButton SelectButton--isSelected');

            await user.click(getButton(1));
            await user.click(getButton(0));

            expect(getButton(0)).not.toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).not.toHaveClass('SelectButton SelectButton--isSelected');
        });

        it('should not allow deletion of last value if field is required', async () => {
            const newProps = {
                isRequired: true,
            };
            rerenderView(newProps);
            const user = userEvent.setup();
            await user.click(getButton(0));

            expect(getButton(0)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).not.toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).not.toHaveClass('SelectButton SelectButton--isSelected');

            await user.click(getButton(1));

            expect(getButton(0)).not.toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).not.toHaveClass('SelectButton SelectButton--isSelected');

            await user.click(getButton(0));

            expect(getButton(0)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).not.toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).not.toHaveClass('SelectButton SelectButton--isSelected');

            await user.click(getButton(1));

            expect(getButton(0)).not.toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(1)).toHaveClass('SelectButton SelectButton--isSelected');
            expect(getButton(2)).not.toHaveClass('SelectButton SelectButton--isSelected');
        });

        it('should call onChange with correct parameters', async () => {
            const user = userEvent.setup();
            await user.click(getButton(1));
            expect(onChangeMock).toHaveBeenLastCalledWith(['1', '2']);
            expect(onChangeMock).toBeCalled();

            await user.click(getButton(2));

            expect(onChangeMock).toBeCalled();
            expect(onChangeMock).toHaveBeenLastCalledWith(['1', '2', '3']);

            await user.click(getButton(1));

            expect(onChangeMock).toBeCalled();
            expect(onChangeMock).toHaveBeenLastCalledWith(['1', '3']);

            await user.click(getButton(0));
            await user.click(getButton(2));

            expect(onChangeMock).toHaveBeenLastCalledWith([]);
        });
    });

    describe('Controlled behavior', () => {
        beforeEach(() => {
            const newProps = {
                defaultValue: undefined,
                value: ['1'],
            };
            rerenderView(newProps);
        });

        it('should render correctly', () => {
            expect(view.container).toMatchSnapshot();
        });

        it('should call onChange with correct parameters', async () => {
            const user = userEvent.setup();

            expect(getButton(0)).toHaveClass('SelectButton SelectButton--isSelected');

            await user.click(getButton(1));

            expect(onChangeMock).toHaveBeenCalled();
        });
    });
});
