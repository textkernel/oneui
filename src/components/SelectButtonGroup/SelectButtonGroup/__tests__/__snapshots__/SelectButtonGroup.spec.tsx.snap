// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectButtonGroup Controlled behavior should render correctly 1`] = `
<div>
  <div
    class="SelectButtonGroup"
  >
    <div
      class="SelectButton SelectButton--isSelected"
      role="button"
      tabindex="0"
    >
      Option 1
    </div>
    <div
      class="SelectButton"
      role="button"
      tabindex="0"
    >
      Option 2
    </div>
    <div
      class="SelectButton"
      role="button"
      tabindex="0"
    >
      Option 3
    </div>
  </div>
</div>
`;

exports[`SelectButtonGroup should render correctly 1`] = `
<div>
  <div
    class="SelectButtonGroup"
  >
    <div
      class="SelectButton SelectButton--isSelected"
      role="button"
      tabindex="0"
    >
      Option 1
    </div>
    <div
      class="SelectButton"
      role="button"
      tabindex="0"
    >
      Option 2
    </div>
    <div
      class="SelectButton"
      role="button"
      tabindex="0"
    >
      Option 3
    </div>
  </div>
</div>
`;
